src/search/mod.rs

```diff
@@ -0,0 +1,96 @@
+use std::collections::HashMap;
+
+use strum::Display;
+
+#[derive(Debug, Display, PartialEq, Eq, Hash)]
+#[strum(serialize_all = "snake_case")]
+pub enum SearchFilter {
+    Album,
+    Artist,
+    Track,
+    Year,
+    Upc,
+    #[strum(serialize = "tag:hipster")]
+    TagHipster,
+    #[strum(serialize = "tag:new")]
+    TagNew,
+    Isrc,
+    Genre,
+}
+
+#[derive(Debug, Default)]
+pub struct SearchQuery {
+    no_filter_query: String,
+    query_map: HashMap<SearchFilter, String>,
+}
```

Comment:
According to the documentation, there is a restriction between filter and type
> You can narrow down your search using field filters. The available filters are `album`, `artist`, `track`, `year`, `upc`, `tag:hipster`, `tag:new`, `isrc`, and `genre`. Each field filter only applies to certain result types.

> The artist and year filters can be used while searching albums, artists and tracks. You can filter on a single year or a range (e.g. 1955-1960).
The `album` filter can be used while searching albums and tracks.
The `genre` filter can be used while searching artists and tracks.
The `isrc` and track filters can be used while searching tracks.
The `upc`, tag:new and tag:hipster filters can only be used while searching albums. The `tag:new` filter will return albums released in the past two weeks, and `tag:hipster` can be used to return only albums with the lowest 10% popularity.

Thus, I think we should take the  `search type` into account when designing the `SearchQuery` Model.

Reply:
I thought about storing the SearchType into the SearchQuery but that means the search method would not need the SearchType anymore and it would be breaking change, I'll try to design this with a method I name differently for to see how it could look

Reply:
I don't see any filters compatible with Shows, Episodes and Playlists. They don't have any advanced filtering keyword ?

Reply:
> 

I don't see any filters compatible with Shows, Episodes and Playlists. They don't have any advanced filtering keyword ?

I think we need to dig deep into the Spotify document to check this, it seems not point out in the Spotify document